File: scattering_angle_calculator.js
class ScatteringAngleCalculator {
    constructor(cellLength = 80, cellWidth = 42, coilLength = 300, coilWidth = 200, detectorWidth = 256) {
        this.cellLength = cellLength;
        this.cellWidth = cellWidth;
        this.coilLength = coilLength;
        this.coilWidth = coilWidth;
        this.detectorWidth = detectorWidth;
    }

    calculateAngles(cellPositionX, sampleWidth) {
        // Calculate scattering angle from the sample (in radians)
        const scatteringangle = Math.atan((this.cellWidth/2 - sampleWidth/2) / (cellPositionX + this.cellLength/2));
        return {
            scatteringangle: scatteringangle
        };
    }

    calculateDetectorWidth(cellPositionX, detectorPositionX, sampleWidth) {
        // サンプル上端→セル右上端→検出器端
        const sampleEdgeX_top = 0;
        const sampleEdgeY_top = sampleWidth / 2;
        const cellEdgeX_top = cellPositionX + this.cellLength / 2;
        const cellEdgeY_top = this.cellWidth / 2;
        const theta_top = Math.atan2(cellEdgeY_top - sampleEdgeY_top, cellEdgeX_top - sampleEdgeX_top);
        const detectorEdgeY_top = sampleEdgeY_top + (detectorPositionX - sampleEdgeX_top) * Math.tan(theta_top);

        // サンプル下端→セル右下端→検出器端
        const sampleEdgeX_bottom = 0;
        const sampleEdgeY_bottom = -sampleWidth / 2;
        const cellEdgeX_bottom = cellPositionX + this.cellLength / 2;
        const cellEdgeY_bottom = -this.cellWidth / 2;
        const theta_2 = Math.atan2(cellEdgeY_bottom - sampleEdgeY_bottom, cellEdgeX_bottom - sampleEdgeX_bottom);
        const detectorEdgeY_2 = sampleEdgeY_bottom + (detectorPositionX - sampleEdgeX_bottom) * Math.tan(theta_2);

        // 検出器での幅
        return Math.abs(detectorEdgeY_top - detectorEdgeY_2);
    }

    plotCoverage(cellPositionX, cellPositionY, detectorPositionX, sampleWidth, showLegend = true) {
        const angles = this.calculateAngles(cellPositionX, sampleWidth);
        const detectorCoverage = this.calculateDetectorWidth(cellPositionX, detectorPositionX, sampleWidth);

        // Sample vertices
        const sampleVertices = [
            [0, -sampleWidth/2],
            [0, sampleWidth/2]
        ];

        // Cell vertices
        const cellVertices = [
            [cellPositionX - this.cellLength/2, cellPositionY - this.cellWidth/2],
            [cellPositionX + this.cellLength/2, cellPositionY - this.cellWidth/2],
            [cellPositionX + this.cellLength/2, cellPositionY + this.cellWidth/2],
            [cellPositionX - this.cellLength/2, cellPositionY + this.cellWidth/2],
            [cellPositionX - this.cellLength/2, cellPositionY - this.cellWidth/2]
        ];

        // Coil vertices
        const coilVertices = [
            [cellPositionX - this.coilLength/2, cellPositionY - this.coilWidth/2],
            [cellPositionX + this.coilLength/2, cellPositionY - this.coilWidth/2],
            [cellPositionX + this.coilLength/2, cellPositionY + this.coilWidth/2],
            [cellPositionX - this.coilLength/2, cellPositionY + this.coilWidth/2],
            [cellPositionX - this.coilLength/2, cellPositionY - this.coilWidth/2]
        ];

        // Detector vertices
        const detectorVertices = [
            [detectorPositionX, 0],
            [detectorPositionX + 1, 0],
            [detectorPositionX + 1, this.detectorWidth/2],
            [detectorPositionX, this.detectorWidth/2],
            [detectorPositionX, 0],
            [detectorPositionX, -this.detectorWidth/2],
            [detectorPositionX + 1, -this.detectorWidth/2],
            [detectorPositionX + 1, 0]
        ];

        // サンプル上端・下端
        const sampleEdgeX = 0;
        const sampleEdgeY = sampleWidth / 2;
        const sampleEdgeY_bottom = -sampleEdgeY;
        // セル右上端・右下端
        const cellEdgeX = cellPositionX + this.cellLength / 2;
        const cellEdgeY = cellPositionY + this.cellWidth / 2;
        const cellEdgeY_bottom = cellPositionY - this.cellWidth / 2;
        // 検出器端
        const detectorEdgeX = detectorPositionX;

        // Scattering Angle 1: サンプル上端→セル右上端→検出器端
        const theta1 = Math.atan2(cellEdgeY - sampleEdgeY, cellEdgeX - sampleEdgeX);
        const detectorEdgeY1 = sampleEdgeY + (detectorEdgeX - sampleEdgeX) * Math.tan(theta1);
        const scatteringLine = {
            x: [sampleEdgeX, cellEdgeX, detectorEdgeX],
            y: [sampleEdgeY, cellEdgeY, detectorEdgeY1],
            type: 'scatter',
            mode: 'lines',
            line: { dash: 'dot', color: 'red', width: 1 },
            name: 'Scattering Angle 1'
        };

        // Scattering Angle 2: サンプル下端→セル右下端→検出器端
        const theta2 = Math.atan2(cellEdgeY_bottom - sampleEdgeY_bottom, cellEdgeX - sampleEdgeX);
        const detectorEdgeY2 = sampleEdgeY_bottom + (detectorEdgeX - sampleEdgeX) * Math.tan(theta2);
        const scatteringLineBottomBottom = {
            x: [sampleEdgeX, cellEdgeX, detectorEdgeX],
            y: [sampleEdgeY_bottom, cellEdgeY_bottom, detectorEdgeY2],
            type: 'scatter',
            mode: 'lines',
            line: { dash: 'dot', color: 'red', width: 1 },
            name: 'Scattering Angle 2'
        };

        // Scattering Angle 3: サンプル上端→セル右下端→検出器端
        const theta3 = Math.atan2(cellEdgeY_bottom - sampleEdgeY, cellEdgeX - sampleEdgeX);
        const detectorEdgeY3 = sampleEdgeY + (detectorEdgeX - sampleEdgeX) * Math.tan(theta3);
        const scatteringLineBottom = {
            x: [sampleEdgeX, cellEdgeX, detectorEdgeX],
            y: [sampleEdgeY, cellEdgeY_bottom, detectorEdgeY3],
            type: 'scatter',
            mode: 'lines',
            line: { dash: 'dot', color: 'black', width: 1 },
            name: 'Scattering Angle 3'
        };

        // Scattering Angle 4: サンプル下端→セル右上端→検出器端
        const theta4 = Math.atan2(cellEdgeY - sampleEdgeY_bottom, cellEdgeX - sampleEdgeX);
        const detectorEdgeY4 = sampleEdgeY_bottom + (detectorEdgeX - sampleEdgeX) * Math.tan(theta4);
        const scatteringLineBottomTop = {
            x: [sampleEdgeX, cellEdgeX, detectorEdgeX],
            y: [sampleEdgeY_bottom, cellEdgeY, detectorEdgeY4],
            type: 'scatter',
            mode: 'lines',
            line: { dash: 'dot', color: 'black', width: 1 },
            name: 'Scattering Angle 4'
        };

        // Create plot data
        const data = [
            {
                x: sampleVertices.map(v => v[0]),
                y: sampleVertices.map(v => v[1]),
                type: 'scatter',
                mode: 'lines',
                line: { color: 'purple', width: 2 },
                name: 'Sample',
                fill: 'toself',
                fillcolor: 'rgba(128,0,128,0.3)'
            },
            {
                x: cellVertices.map(v => v[0]),
                y: cellVertices.map(v => v[1]),
                type: 'scatter',
                mode: 'lines',
                line: { color: 'blue', width: 2 },
                name: 'Cell',
                fill: 'none'
            },
            {
                x: coilVertices.map(v => v[0]),
                y: coilVertices.map(v => v[1]),
                type: 'scatter',
                mode: 'lines',
                line: { color: 'green', width: 2 },
                name: 'Coil',
                fill: 'none'
            },
            {
                x: detectorVertices.map(v => v[0]),
                y: detectorVertices.map(v => v[1]),
                type: 'scatter',
                mode: 'lines',
                line: { color: 'red', width: 2 },
                name: 'Detector',
                fill: 'toself',
                fillcolor: 'rgba(255,0,0,0.3)'
            },
            scatteringLine,
            scatteringLineBottomBottom,
            scatteringLineBottom,
            scatteringLineBottomTop
        ];

        // Layout settings
        const layout = {
            title: {
                text: '',
                x: 0.02,
                y: 0.3,
                xanchor: 'left',
                yanchor: 'top'
            },
            xaxis: {
                title: 'X (mm)',
                range: [-50, detectorPositionX + 50],
                zeroline: true,
                gridcolor: 'lightgray'
            },
            yaxis: {
                title: 'Y (mm)',
                range: [-Math.max(this.detectorWidth/2, sampleWidth) - 50, Math.max(this.detectorWidth/2, sampleWidth) + 50],
                zeroline: true,
                gridcolor: 'lightgray'
            },
            showlegend: showLegend,
            legend: {
                x: 1,
                y: 1,
                xanchor: 'right',
                yanchor: 'top',
                draggable: true
            },
            margin: { t: 50, r: 120 },
            annotations: [
                {
                    x: 0,
                    y: Math.max(this.detectorWidth/2, sampleWidth) + 30,
                    xref: 'x',
                    yref: 'y',
                    text: `${cellPositionY > 0 ? 'Scattering Angle 1' : cellPositionY === 0 ? 'Scattering Angle' : 'Scattering Angle 2'}: ${cellPositionY === 0 ? '±' : ''}${((cellPositionY >= 0 ? theta1 : theta2) * 180 / Math.PI).toFixed(1)}°<br>Detector Coverage: ±${detectorCoverage.toFixed(1)/2}mm`,
                    showarrow: false,
                    font: {
                        size: 14,
                        color: 'black'
                    },
                    bgcolor: 'rgba(255, 255, 255, 0.95)',
                    bordercolor: 'black',
                    borderwidth: 1,
                    borderpad: 4,
                    xanchor: 'left',
                    yanchor: 'top'
                }
            ]
        };

        // Draw plot
        const config = {
            editable: true,
            dragmode: 'pan'
        };
        Plotly.newPlot('plot', data, layout, config);
    }
}

// Set up event listeners
function setupEventListeners() {
    const inputs = document.querySelectorAll('input');
    inputs.forEach(input => {
        input.addEventListener('input', updatePlot);
        input.addEventListener('change', updatePlot);
    });
}

// Update plot
function updatePlot() {
    const calculator = new ScatteringAngleCalculator(
        parseFloat(document.getElementById('cellLength').value),
        parseFloat(document.getElementById('cellWidth').value),
        parseFloat(document.getElementById('coilLength').value),
        parseFloat(document.getElementById('coilWidth').value),
        parseFloat(document.getElementById('detectorWidth').value)
    );

    const sampleWidth = parseFloat(document.getElementById('sampleWidth').value);
    const cellPosition = parseFloat(document.getElementById('cellPosition').value);
    const cellPositionY = parseFloat(document.getElementById('cellPositionY').value);
    const detectorPosition = parseFloat(document.getElementById('detectorPosition').value);
    const showLegend = document.getElementById('showLegend').checked;

    calculator.plotCoverage(cellPosition, cellPositionY, detectorPosition, sampleWidth, showLegend);
}

// Initialize
document.addEventListener('DOMContentLoaded', () => {
    setupEventListeners();
    updatePlot();
});

File: scattering_angle_calculator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
散乱角計算プログラム
ガラスセル、コイル、検出器の配置から散乱角と検出幅を計算・可視化します。
"""

import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl

# 日本語フォントの設定
plt.rcParams['font.family'] = 'Hiragino Sans'  # macOSの標準日本語フォント
mpl.rcParams['axes.unicode_minus'] = False  # マイナス記号を正しく表示

class ScatteringAngleCalculator:
    def __init__(self, cell_length=80, cell_width=42, coil_length=300, coil_width=200, detector_width=256):
        """
        散乱角計算器のコンストラクタ

        Parameters:
        -----------
        cell_length : float
            ガラスセルの長さ (mm)
        cell_width : float
            ガラスセルの幅 (mm)
        coil_length : float
            コイルの長さ (mm)
        coil_width : float
            コイルの幅 (mm)
        detector_width : float
            検出器の幅 (mm)
        """
        self.cell_length = cell_length
        self.cell_width = cell_width
        self.coil_length = coil_length
        self.coil_width = coil_width
        self.detector_width = detector_width

    def calculate_angles(self, cell_position_x):
        """
        与えられたガラスセルの位置における散乱角を計算します。

        Parameters:
        -----------
        cell_position_x : float
            試料からガラスセルの中心までのX軸方向の距離 (mm)

        Returns:
        --------
        float
            散乱角（ラジアン）
        """
        # 散乱角の計算（ラジアンで返す）
        scattering_angle = np.arctan((self.cell_width/2) / (cell_position_x+self.cell_length/2))
        return scattering_angle

    def calculate_detector_width(self, cell_position_x, detector_position_x):
        """
        散乱角によって作られる検出器上の幅を計算します。

        Parameters:
        -----------
        cell_position_x : float
            試料からガラスセルの中心までのX軸方向の距離 (mm)
        detector_position_x : float
            試料から検出器までのX軸方向の距離 (mm)

        Returns:
        --------
        float
            検出器上の幅 (mm)
        """
        scattering_angle = self.calculate_angles(cell_position_x)

        # 散乱角（ラジアン）のtanを使って検出器位置でのy座標を計算
        detector_y = detector_position_x * np.tan(scattering_angle)

        # 検出器上の幅を返す（中心から片側の距離の2倍）
        return detector_y * 2

    def plot_coverage(self, cell_position_x, detector_position_x=None):
        """
        散乱角の範囲を可視化します。

        Parameters:
        -----------
        cell_position_x : float
            試料からガラスセルの中心までのX軸方向の距離 (mm)
        detector_position_x : float, optional
            試料から検出器までのX軸方向の距離 (mm)
        """
        scattering_angle = self.calculate_angles(cell_position_x)

        # プロット作成
        fig, ax = plt.subplots(figsize=(12, 8))

        # ガラスセルの範囲をプロット（x軸上の指定位置に、y≥0の部分のみ）
        rect_cell = plt.Rectangle((cell_position_x - self.cell_length/2, 0),
                            self.cell_length, self.cell_width/2,
                            fill=False, color='blue', label='ガラスセル', linewidth=1.5)
        ax.add_patch(rect_cell)

        # コイルの範囲をプロット（x軸上の指定位置に、y≥0の部分のみ）
        rect_coil = plt.Rectangle((cell_position_x - self.coil_length/2, 0),
                            self.coil_length, self.coil_width/2,
                            fill=False, color='green', label='コイル', linewidth=1.5)
        ax.add_patch(rect_coil)

        # 検出器をプロット（指定された場合）
        if detector_position_x is not None:
            detector = plt.Rectangle((detector_position_x, 0),
                                  1, self.detector_width/2,
                                  fill=True, color='red', alpha=0.3, label='検出器')
            ax.add_patch(detector)

            # 散乱角による検出器上の幅を計算
            detector_coverage = self.calculate_detector_width(cell_position_x, detector_position_x)

        # 原点から四隅への線をプロット（y≥0の部分のみ）
        corners = [
            (cell_position_x + self.cell_length/2, 0),
            (cell_position_x + self.cell_length/2, self.cell_width/2),
        ]

        # 散乱角の線を描画
        for x, y in corners:
            if detector_position_x is not None:
                # 検出器位置までの線を引く
                y_at_detector = (y / x) * detector_position_x
                if y_at_detector >= 0:  # y≥0の場合のみプロット
                    plt.plot([0, detector_position_x], [0, y_at_detector], 'k--', alpha=0.3)
            else:
                # 検出器がない場合は四隅まで
                if y >= 0:  # y≥0の場合のみプロット
                    plt.plot([0, x], [0, y], 'k--', alpha=0.3)

        # グラフの設定
        ax.set_aspect('equal')
        plt.grid(True)

        # タイトルをグラフ内部に表示
        text_content = f'散乱角: {np.degrees(scattering_angle):.1f}°'
        if detector_position_x is not None:
            text_content += f'\n検出幅: {detector_coverage:.1f}mm'

        # グラフの上部内側にテキストを配置
        plt.text(0.02, 0.98, text_content,
                transform=ax.transAxes,  # 軸に対する相対座標を使用
                verticalalignment='top',
                horizontalalignment='left',
                bbox=dict(facecolor='white', alpha=0.8, edgecolor='none'))

        plt.xlabel('X (mm)')
        plt.ylabel('Y (mm)')
        plt.legend()

        # 軸の範囲設定
        margin = 50
        max_x = max(cell_position_x + self.cell_length/2, detector_position_x if detector_position_x else 0)
        plt.xlim(-margin, max_x + margin)
        plt.ylim(0, self.detector_width/2 + margin)  # y軸は0以上のみ表示

        plt.show()

def main():
    """メイン関数"""
    # パラメータ設定
    cell_position = 200    # ガラスセルの位置 (mm)
    detector_position = 800  # 検出器の位置 (mm)

    # 計算実行
    calculator = ScatteringAngleCalculator()
    calculator.plot_coverage(cell_position, detector_position)

if __name__ == "__main__":
    main()
